Introduce `gh agent-task list`

BagToad opened the pull request:
This pull request introduces a new `agent-task` command group with a `list` subcommand for interacting with GitHub Copilot agent tasks, including listing in-flight agent sessions and hydrating them with pull request data. It also adds supporting infrastructure for Copilot API communication, updates the API models, and includes comprehensive tests for the new functionality.

BagToad committed:
Add agent task listing command and CAPI client

Introduces a new 'list' subcommand under agent-task for listing agent tasks. Implements a Copilot API client for fetching agent sessions and hydrating them with pull request data. Updates PullRequest and PRRepository types to support new fields. Adds dependencies for msgpack and tagparser.

Co-Authored-By: Babak K. Shandiz <babakks@github.com>

babakks commented on pkg/cmd/agent-task/list/list.go:
```diff
@@ -0,0 +1,150 @@
+			}
+
+			authCfg := cfg.Authentication()
+			opts.CapiClient = capi.NewCAPIClient(httpClient, authCfg)
```
I know why you did this in `RunE`, but I'd rather keep the code consistent with other commands.

How about adding a `CapiHTTPClient` field with the same signature as `HttpClient` to the opts struct, and assign it when we're registering the flags (i.e. outside of, and chronologically before, `RunE`). Then we can do HTTP stubbing as usual by overriding `CapiHTTPClient`.

We can even remove the `HttpClient` field from opts.

BagToad replied:
Fixed in 7b71b5f21b710834dbca1257c0062c7437c3d9c0

babakks commented on pkg/cmd/agent-task/list/list.go:
```diff
@@ -0,0 +1,150 @@
+		} else {
+			// Skip these sessions in case they happen, for now.
+			continue
+		}
```
How about simplifying these `if`-statements like this?

```suggestion
		if s.ResourceType != "pull" || s.PullRequest == nil || s.PullRequest.Repository == nil {
			// Skip these sessions in case they happen, for now.
			continue
		}

		pr := fmt.Sprintf("#%d", s.PullRequest.Number)
		repo := s.PullRequest.Repository.NameWithOwner
```

I think it's a legitimate expectation to have non-zero/non-empty fields on `s.PullRequest` or `s.PullRequest.NameWithOwner`, and we should only check if they're not `nil` (just for the sake of language semantics).

BTW, if you want to keep it as is, I think we should at least add `s.PullRequest != nil` to both `if`s.

BagToad replied:
Thank you, I like it. Done in 0a5b78a510a7ed44ea833fb3e3da784e0243287d

babakks commented on pkg/cmd/agent-task/list/list.go:
```diff
@@ -0,0 +1,150 @@
+				stateColor = cs.Red
+			default:
+				stateColor = cs.Muted
+			}
```
This can be extracted as a shared function for reuse.

BagToad replied:
Done in 4fd6ae6e3a980a3911ee32c6e2be9e68488d53a9. I also introduced a shared package for this since we'll need it across the whole command set.

babakks commented on pkg/cmd/agent-task/capi/sessions.go:
```diff
@@ -0,0 +1,208 @@
+
+	apiClient := api.NewClientFromHTTP(c.httpClient)
+
+	var resp struct {
+		Nodes []struct {
+			PullRequest sessionPullRequest `graphql:"... on PullRequest"`
+		} `graphql:"nodes(ids: $ids)"`
+	}
```
I guess we're fine as long as we don't expose this intermediate type, `sessionPullRequest`.

FWIW I tried to override the field names like this but it didn't work (got a query compilation error). 

```
PullRequest sessionPullRequest `graphql:"... on PullRequest {id,fullDatabaseID,number,title,state,url,body,createdAt,updatedAt,closedAt,mergedAt,repository {id,name,nameWithOwner}}"`
```

This is the error:

```
GraphQL: Expected NAME, actual: LCURLY ("{") at [1, 69]
```

So, yeah, let's just move on.

babakks commented on pkg/cmd/agent-task/capi/sessions.go:
```diff
@@ -0,0 +1,208 @@
+		return nil, err
+	}
+
+	return Sessions, nil
```
Var name is capitalised.


```suggestion
	result, err := c.hydrateSessionPullRequests(sessions)
	if err != nil {
		return nil, err
	}

	return result, nil
```

BagToad replied:
Fixed in 13c293f4cfb08f8af46936d5d3469969005b1fa8

babakks commented on pkg/cmd/agent-task/capi/sessions.go:
```diff
@@ -0,0 +1,208 @@
+	encoded := base64.RawURLEncoding.EncodeToString(buf.Bytes())
+
+	// Return with the PR_ prefix
+	return "PR_" + encoded
```
We don't need the comments since the code is clear enough.


```suggestion
	if err := encoder.Encode(parts); err != nil {
		panic(err)
	}

	encoded := base64.RawURLEncoding.EncodeToString(buf.Bytes())
	return "PR_" + encoded
```

BagToad replied:
Fixed in 5281be467d5ab084e716b3c96fedc6a698013438

babakks commented on pkg/cmd/agent-task/capi/sessions.go:
```diff
@@ -0,0 +1,208 @@
+	}
+
+	host, _ := c.authCfg.DefaultHost()
+	err := apiClient.Query(host, "FetchPRs", &resp, map[string]any{
```
Maybe a more specific query name that shows this is related to fetching session PRs? like `FetchPRsForAgentTaskSessions`?

BagToad replied:
Fixed in bc1d306c31111ca9ca371265f6e292bc8ba5e14d

babakks commented on pkg/cmd/agent-task/capi/sessions.go:
```diff
@@ -0,0 +1,208 @@
+				})
+			}
+		}
+	}
```
How about using a map to make it $O(n)$?


```suggestion
	prMap := make(map[string]*api.PullRequest, len(prNodeIds))
	for _, node := range resp.Nodes {
		prMap[node.PullRequest.FullDatabaseID] = &api.PullRequest{
			ID:             node.PullRequest.ID,
			FullDatabaseID: node.PullRequest.FullDatabaseID,
			Number:         node.PullRequest.Number,
			Title:          node.PullRequest.Title,
			State:          node.PullRequest.State,
			URL:            node.PullRequest.URL,
			Body:           node.PullRequest.Body,
			CreatedAt:      node.PullRequest.CreatedAt,
			UpdatedAt:      node.PullRequest.UpdatedAt,
			ClosedAt:       node.PullRequest.ClosedAt,
			MergedAt:       node.PullRequest.MergedAt,
			Repository:     node.PullRequest.Repository,
		}
	}

	newSessions := make([]*Session, 0, len(sessions))
	for _, s := range sessions {
		newSessions = append(newSessions, &Session{
			session:     s,
			PullRequest: prMap[strconv.FormatInt(s.ResourceID, 10)],
		})
	}
```

BagToad replied:
Fixed in ea9dfae3bf16b6730e4d5e23e6ca5caefd8b7c68

babakks reviewed (commented):
Haven't yet checked the tests, but, overally, it looks very good! 🙏 

Just had a couple of comments.

babakks commented on pkg/cmd/agent-task/list/list.go:
```diff
@@ -0,0 +1,150 @@
+	"github.com/cli/cli/v2/internal/gh"
+	"github.com/cli/cli/v2/internal/tableprinter"
+	"github.com/cli/cli/v2/pkg/cmd/agent-task/capi"
+	"github.com/cli/cli/v2/pkg/cmd/pr/shared"
```
How about `prShared` as a little remark that this comes from another command set?

```suggestion
	prShared "github.com/cli/cli/v2/pkg/cmd/pr/shared"
```

If you decided to apply this then you should also change L107 as well.

BagToad replied:
Done in 0de5cf24f070fd6849712a95a8267e8ce00dc94a

babakks commented on pkg/cmd/agent-task/capi/sessions.go:
```diff
@@ -0,0 +1,208 @@
+	// Uncomment one of these to see error
+	// Author api.Author
+	// MergedBy   *api.Author
+	Repository *api.PRRepository
```
Yeah, you're right. Already shared my thoughts, so we can just remove the comments here.

BagToad replied:
Fixed in 48012063df33955523c064681599e9acd97b710b

babakks commented on pkg/cmd/agent-task/capi/sessions.go:
```diff
@@ -0,0 +1,208 @@
+
+	var sessions []session
+	page := 1
+	perPage := 50
```
FWIW the API default for this is 20.

babakks reviewed (commented):
A couple more notes.

BagToad committed:
Fix import alias for shared package in agent-task list

Renames the import of the shared package to 'prShared' and updates its usage in list.go to avoid naming conflicts and improve code clarity.

BagToad committed:
Refactor session filtering in listRun function

Simplifies logic for filtering sessions to only include those with valid pull request and repository data. This reduces nested conditionals and improves code readability.

BagToad committed:
Refactor agent-task list command client initialization

Moves CAPI client initialization to a deferred function in ListOptions, simplifying command setup and improving testability. Updates tests to use the new client initialization pattern and adds more comprehensive test cases for session listing.

BagToad committed:
Remove commented-out fields from sessionPullRequest

Cleaned up the sessionPullRequest struct by deleting unused commented-out fields related to Author and MergedBy.

BagToad committed:
Refactor variable names in ListSessionsForViewer

Renamed 'Sessions' to 'result' for clarity and consistency in the ListSessionsForViewer method when hydrating session pull requests.

BagToad committed:
Update API query name for session PR fetch

Changed the API query from 'FetchPRs' to 'FetchPRsForAgentTaskSessions' in hydrateSessionPullRequests to match updated backend endpoint.

BagToad committed:
Refactor error handling in generatePullRequestNodeID

Simplified error handling by inlining the encoder.Encode call and removing redundant comments for improved code clarity.

BagToad committed:
Optimize session pull request hydration logic

Replaces linear search with a map for associating sessions with pull requests, improving performance and simplifying code in hydrateSessionPullRequests.

BagToad committed:
Refactor session state color logic to shared package

Moved the session state color selection logic from list.go to a new shared/display.go file as ColorFuncForSessionState. This improves code reuse and maintainability by centralizing the color mapping for session states.
